### DISCUSSION TITLE
Extract embedding from my model for other use.

### DISCUSSION THREAD

**Discussion by jjkai** [ludwig-ai/ludwig#3393] [Q&A]
Hi I have question on how can I extract the embedding from my model for my other downstream tasks. Say for example, running k-means clustering. 

Context: I am training a text classification model following the example here: https://ludwig.ai/latest/examples/text_classification/ 
However, instead of getting the prediction result for each of the records, I am also interested in extracting the embedding for each record: the output from the last fully connected layer before making the prediction. That way I can try to run some other analysis to see what information the model is capturing. As simple as running a k-means clustering to see if each cluster contains similar data records. 

Any advise and feedback are appreciated. Thank you.

**Comment by tgaddair** [comment]
Hey @jjkai, good question. This is possible in Ludwig using the CLI command `ludwig collect_activations` as described [here](https://ludwig.ai/latest/user_guide/command_line_interface/#collect_activations). Unfortunately we don't have a lot of great examples of this (we should add some!) but we do have an example test here that demonstrates usage:

https://github.com/ludwig-ai/ludwig/blob/master/tests/integration_tests/test_collect.py#L91

Let me know if you have any follow-up questions or suggestions for how we can improve the docs / features around this!

**Reply by jjkai** [reply]
Thank you! I will give it a try and let you know if I have further questions/suggestions.

**Reply by jjkai** [reply]
Hi @tgaddair I tried to run the `collect_activations` function and it generates three files: 'tag__last_hidden.npy', 'tag__logits.npy', and 'tag__projection_input.npy', which I believe is the expected behavior. So the 'tag__logits.npy' looks pretty straightforward to me should be the final predicted logits of an input. But I am not super certain about 'tag__last_hidden.npy' and 'tag__projection_input.npy'. I am guessing 'tag__last_hidden.npy' is probably the node weights for the last layer? And 'tag__projection_input.npy' probably is the embedding I am looking for? But this is all my speculation, can you please verify or provide more detail? 

One other issue I am encountering is that when I run the `collect_activations` function with a dataset say 200 records, sometimes the returned files 'tag__last_hidden.npy' and 'tag__projection_input.npy' would contain less than 200 records. This is likely the sign that I am done something incorrectly. In addition, when looking into the values of the returned records (each is a vector with 256 values), I found that the majority of them are 0. I am attaching an image below. 
![image](https://github.com/ludwig-ai/ludwig/assets/8563823/b7ecd518-0bf8-47e0-9d3c-cc6f01db764f)

Not exactly sure what went wrong in the process. Maybe as you suggested, a more concrete example would be great and super helpful. But in the meantime, I appreciate if you can share some insights or maybe know what the issue is already. 

Thanks!
